### DISCUSSION TITLE
[1.1.448] Removal of unit preset feature

### DISCUSSION THREAD

**Discussion by huchenlei** [Mikubill/sd-webui-controlnet#2854] [General]
Unit preset feature is removed in 1.1.448 mainly because it is becoming hard to maintain and buggy to use. The core problem comes from gradio's event system. There is no way to update UI state without triggering subsequent state updates. For example, if you update control type and preprocessor in a single callback, control type change will further trigger the preprocessor and model to change, which will overwrite the preprocessor set initially.

I also found that we are listening for excessive amount of gradio events to reconstruct ControlNetUnit state. Part of this is also caused by preset. This has slowed the UI response. 1.1.448 has also removed these excessive listeners.

If you think the preset feature should come back later. Please leave a comment on how you think this should be achieved. Personally I think preset probably need to be a pure front-end solution to achieve the things we want.

There are a few other breaking changes regarding use of enum and preprocessor slider default values. You can read details here: https://github.com/Mikubill/sd-webui-controlnet/pull/2847.

**Comment by sdbds** [comment]
I think the preset part is better managed by an external plugin from loading config, such as [state](https://github.com/ilian6806/stable-diffusion-webui-state) or [lightdiffusionflow](https://github.com/Tencent/LightDiffusionFlow)

**Reply by light-and-ray** [reply]
I don't agree, in ControlNet I used presets for combinations which are not in control param defaults

**Reply by sdbds** [reply]
That means we should add more control params.

**Reply by light-and-ray** [reply]
I had these methods which are not in quick assess:
1. T2I Color. Also Sketch, Style can be separated or moved into similar Lineart and Reference
2. Densepose. It's too different to be in the same section with OpenPose.
3. QR code pattern. (for hidden text etc) It doesn't have preprocessor, only "none", "invert", but I think it's enough popular to have dedicated "Control Type"
4. For SparseCtrl default must be "none" + "mm_sd15_v3_sparsectrl_rgb", but there is mixed up default "scribble_pidinet" + "mm_sd15_v3_sparsectrl_rgb" which doesn't make sense

**Comment by light-and-ray** [comment]
I think it can be achieved by on_after_element callbacks, which automatically registers all components with specific element ids. I think it will be very stable, but idk about ui lags.

I'm planning to do inside my extension [replacer](https://github.com/light-and-ray/sd-webui-replacer), and after that, if it is good, I will make a PR here with the same way. If it is laggy, I will make it optional

**Comment by Surran-Peng** [comment]
something wrong with 1.1.448

the mode "adetailer + controlnet + x/y/z plot" worked fine before v447. but now it generates difference images after the second image.

example:
step 1:Use "controlnet + x/y/z plot" to find the sets of Steps and CFG value.
step 2:Add "adetailer" to get a good image. Except for the first image, the others will be different. not the one found by step 1.

**Reply by huchenlei** [reply]
Please file an issue for tracking. Thanks! 1.1.448 does a very large refactoring on ControlNetUnit, and our test case are not including how the extension works with other extensions, so something might break.

**Reply by Surran-Peng** [reply]
Thank you for your reply.

I'm new to github and don't know how to file an issue.

Anyway, can you tell me how to get back to v447?

**Reply by light-and-ray** [reply]
Just click "..." and select "reference in new issue"

**Comment by RedDestiny** [comment]
我特别喜欢依赖这个功能，希望能维护这个预设的功能啊！
